// plugins/HistoryPlugin.js
var O2 = require('o2');
var Mediator = require('mediator');
var fs = require('fs');
var HISTORY_STORAGE = 'history.storage';
var HistoryPlugin = O2.extendClass(Mediator.Plugin, {

   date : false,
   fileLog : false,

   getName : function() {
      return 'History';
   },

   init : function() {
      this.register('message');
      var d = new Date();
      this.date = d.getFullYear() + "-" + d.getMonth() + "-" + d.getDay();
      this.fileLog = HISTORY_STORAGE + "/" + this.date + ".log";
      try {
         fs.statSync(HISTORY_STORAGE);
      } catch (e) {
         fs.mkdirSync(HISTORY_STORAGE);
      }
   },

   message : function(ctx) {
      var oUser = ctx.user;
      var sName = oUser.getName();
      var sMsg = ctx.message;
      var d = new Date();
      var time = d.getHours() + ":" + d.getMinutes();
      var sLog = time + ":" + sName + ":" + sMsg + "\n";
      fs.appendFileSync(this.fileLog, sLog);
      ctx.history = {
         user : oUser.getName()
      };
   },
   

});

module.exports = HistoryPlugin;
