// plugins/LogPlugin.js
var O2 = require('o2');
var Mediator = require('mediator');

var LogPlugin = O2.extendClass(Mediator.Plugin, {
	
	getName: function() {
		return 'Log';
	},
	
	init: function() {
		this.register('channelCreated');
		this.register('channelDestroyed');
		this.register('userConnected');
		this.register('userDisconnected');
		this.register('userJoined');
		this.register('userLeft');
		this.register('log');
		// this.register('message');
	},
	
	log: function() {
		var d = new Date();
		var s = Array.prototype.slice.call(arguments, 0).join(' ');
		console.log('[' + d.toJSON() + '] [T] ' + s);
	},
	
	userConnected: function(oUser) {
		this.log('user "' + oUser.sName + '" connected');
	},
	
	userDisconnected: function(oUser) {
		this.log('user "' + oUser.sName + '" disconnected');
	},
	
	channelCreated: function(oChannel) {
		this.log('channel "' + oChannel.sName + '" created');
	},
	
	channelDestroyed: function(oChannel) {
		this.log('channel "' + oChannel.sName + '" destroyed');
	},
	
	userJoined: function(oChannel, oUser) {
		this.log('user "' + oUser.sName + '" has joined channel "' + oChannel.sName  + '"');
	},
	
	userLeft: function(oChannel, oUser) {
		this.log('user "' + oUser.sName + '" has left channel "' + oChannel.sName + '"');
	},
	
	message: function(oMessageContext) {
		var sMessage = oMessageContext.message, 
			oChannel = oMessageContext.channel, 
			oUser = oMessageContext.user;
		var sChannelName = '';
		if (oChannel) {
			sChannelName = oChannel.sName;
		} else {
			sChannelName = '@' + oUser.sName;
		}
		this.log(oUser.sName + ' (' + sChannelName + '): ' + sMessage);
	}
	
});

module.exports = LogPlugin;
