// txat/index.js
var wsh = require('wshelper');
var http = require('httphelper');
var TxatSystem = require('./System.js');
var oTxat;
var aSocketRegistry = [];

var CONST = require('./consts.js');
var util = require('util');
var fs = require('fs');
var Password = require('./Password.js');

////// TXAT SYSTEM ////// TXAT SYSTEM ////// TXAT SYSTEM ////// TXAT SYSTEM //////
////// TXAT SYSTEM ////// TXAT SYSTEM ////// TXAT SYSTEM ////// TXAT SYSTEM //////
////// TXAT SYSTEM ////// TXAT SYSTEM ////// TXAT SYSTEM ////// TXAT SYSTEM //////


function serverLog() {
	var a = Array.prototype.slice.call(arguments, 0);
	http.writeLog('[T] ' + a.join(' '));
}

function txatInit(c) {
	oTxat = new TxatSystem();
	oTxat.onMessage = txatMessage;
	oTxat.onNotify = txatNotify;
	oTxat.configure(c);

	// Les addons
	var sPath = 'www/txat/scripts/Txat/addons';
	// au démarrage un sync conviendra
	var aAddons = fs.readdirSync(sPath);
	var aInclude = [
		'<base href="/txat/" />'
	];
	aAddons.forEach(function(sAddon) {
		try {
			var sDirectory = sPath + '/' + sAddon;
			var oStat = fs.statSync(sDirectory);
			aInclude.push(fs.readFileSync(sDirectory + '/index.html'));
		} catch(e) {
			// probablement pas d'index
		}
	}, this);
	
	var sIncludeAddons = aInclude.join('\n');
	
	http.defineCommand('txat-addons', function(response, query) {
		response.setHeader('Content-Type', 'text/plain');
		response.end(sIncludeAddons);
	});
}

/**
 * Function appelée en callback lorsqu'un client se connecte
 * @param Socket oSocket
 */

function txatService(oSocket) {
	var sAddress = wsh.getAddress(oSocket);
	oSocket.on('disconnect', function() {
		if (wsh.checkSession(oSocket)) {
			// signaler la déconnexion aux autres
			var nId = wsh.getData(oSocket, 'txat_id');
			if (!!nId) {
				serverLog(CONST.MESSAGE.DISCONNECTED, sAddress);
				// exclure le client du systeme de chat				
				oTxat.dropUser(nId);
				aSocketRegistry[nId] = null;
			}
		} else {
			serverLog('ignoring previous session socket disconnection');
		}
	});
	
	////// MESSAGES TXAT ////// MESSAGES TXAT ////// MESSAGES TXAT ////// MESSAGES TXAT //////
	////// MESSAGES TXAT ////// MESSAGES TXAT ////// MESSAGES TXAT ////// MESSAGES TXAT //////
	////// MESSAGES TXAT ////// MESSAGES TXAT ////// MESSAGES TXAT ////// MESSAGES TXAT //////
	
	/**
	 * demande d'authentification
	 * T_LOGIN 
	 * - u: string, pseudonyme soumit par l'utilisateur
	 */
	oSocket.on(CONST.OPCODE.C_LOGIN, function(xData) {
		try {
			var sUserName = xData.u;
			var sPass;
			if ('p' in xData) {
				sPass = xData.p;
			}
			oTxat.checkUserLoginName(sUserName);
			Password.check(sUserName, sPass).then(function() {
				serverLog('access granted for user', sUserName);
				var u = oTxat.createUser(sUserName);
				wsh.setData(oSocket, 'txat_id', u.id);
				aSocketRegistry[u.id] = oSocket;
				wsh.send([oSocket], CONST.OPCODE.HI, {u: sUserName});
				oTxat.userJoinsChannel(u.id, oTxat.getChannel(1).sName);
			}, function(err) {
				serverLog('access denied for user', sUserName, ': invalid password');
				oSocket.emit(CONST.OPCODE.DN, {e: err.toString()});
			});
		} catch (e) {
			console.log(e.stack);
			wsh.send([oSocket], CONST.OPCODE.DN, {e: e.toString()});
		}
	});
	
	/** 
	 * message de discussion public
	 * T_SAY
	 * - m: string, contenu du message.
	 * - c: string, référence du canal sur lequel envoyer le message
	 */
	oSocket.on(CONST.OPCODE.C_SAY, function(xData) {
		try {
			if (xData.m !== '') {
				var nId = wsh.getData(oSocket, 'txat_id');
				oTxat.sendMessageToChannel(nId, oTxat.getChannelId(xData.c), xData.m);
			}
		} catch (e) {
			txatErrorMessage(oSocket, e);
		}
	});
	
	/**
	 * message privé à destination d'un utilisateur
	 * T_MSG
	 * - m: string, contenu du message
	 * - u: nom de l'utilisateur destinataire
	 */
	oSocket.on(CONST.OPCODE.C_MSG, function(xData) {
		try {
			var nId = wsh.getData(oSocket, 'txat_id');
			oTxat.sendMessageToUser(nId, oTxat.getUserId(xData.u), xData.m);
		} catch (e) {
			txatErrorMessage(oSocket, e);
		}
	});

	/**
	 * rejoindre un canal
	 * T_JOIN
	 * - c: string, nom du canal à rejoindre
	 */
	oSocket.on(CONST.OPCODE.C_JOIN, function(xData) {
		try {
			var nId = wsh.getData(oSocket, 'txat_id');
			oTxat.userJoinsChannel(nId, xData.c);
		} catch (e) {
			txatErrorMessage(oSocket, e);
		}
	});

	/**
	 * quitter un canal
	 * T_LEAVE
	 * - c: nom du canal à quitter
	 */
	oSocket.on(CONST.OPCODE.C_LEAVE, function(xData) {
		try {
			var nId = wsh.getData(oSocket, 'txat_id');
			oTxat.removeUserFromChannel(nId, oTxat.getChannelId(xData.c));
		} catch (e) {
			txatErrorMessage(oSocket, e);
		}
	});
	
	/**
	 * Demander la liste des canaux qui ont été créé sur le serveur
	 * et auxquels l'utilisateur à le droit de se connecter
	 * T_LIST
	 * 
	 */
	oSocket.on(CONST.OPCODE.C_LIST, function(xData) {
		var nId = wsh.getData(oSocket, 'txat_id');
		var aList = oTxat.getUserAccessibleChannels(nId);
		oSocket.emit(CONST.OPCODE.LS, {c: aList});
	});
	
	/**
	 * Demander des informations sur un utilisateur.
	 * Les information sont relative à un canal données
	 * Ce qui fait que cette commande permet de découvrir le profil 
	 * utilisateur dans le contexte du canal spécifié.
	 * 
	 * T_WHO
	 * - u: string, nom de l'utilisateur
	 * - c: string, canal de discussion pour déterminer les droits de cet utilisateur dans ce canal.
	 */
	oSocket.on(CONST.OPCODE.C_WHO, function(xData) {
		try {
			var sWho = xData.u;
			var sChannel = xData.c;
			var idChannel = oTxat.getChannelId(sChannel);
			var oUser = oTxat.getUser(oTxat.getUserId(sWho));
			var oChannel = oTxat.getChannel(idChannel);
			if (idChannel.toString() in oUser.oPowers) {
				var p = oUser.oPowers[idChannel.toString()];
				var sInfo = util.format(CONST.MESSAGE.USER_ON_CHANNEL, sWho, p.getRankStr(), xData.c);
				oSocket.emit(CONST.OPCODE.IM, {c: xData.c, m: sInfo});
			} else {
				txatErrorMessage(oSocket, util.format(CONST.MESSAGE.USER_NOT_ON_CHANNEL, xData.u, xData.c));
			}
		} catch (e) {
			txatErrorMessage(oSocket, e);
		}
	});
	
	oSocket.on(CONST.OPCODE.C_BAN, function(xData) {
		try {
			var nId = wsh.getData(oSocket, 'txat_id');
			var sChannel = xData.c;
			if ('u' in xData) {
				// ban user
				var sUser = xData.u;
				var nTime = xData.t | 0;
				var sWhy = xData.m;
				oTxat.banUserFromChannel(oTxat.getUserId(sUser), oTxat.getChannelId(sChannel), nTime, sWhy, nId);
			} else {
				// ban reports
				var oChannel = oTxat.getChannel(oTxat.getChannelId(sChannel));
				var oReport = oChannel.getBanReports(' - ');
				var aTo = [oSocket];
				var r, sReport;
				for (var sUser in oReport) {
					wsh.send(aTo, CONST.OPCODE.IM, {c: sChannel, m: sUser + oReport[sUser]});
				}
			}
		} catch (e) {
			txatErrorMessage(oSocket, e);
		}
	});

	oSocket.on(CONST.OPCODE.C_UNBAN, function(xData) {
		try {
			var nId = wsh.getData(oSocket, 'txat_id');
			var sUser = xData.u;
			var sChannel = xData.c;
			oTxat.unbanUserFromChannel(oTxat.getUserId(sUser), oTxat.getChannelId(sChannel), nId);
		} catch (e) {
			txatErrorMessage(oSocket, e);
		}
	});

	oSocket.on(CONST.OPCODE.C_PROMOTE, function(xData) {
		try {
			var nId = wsh.getData(oSocket, 'txat_id');
			var sUser = xData.u;
			var sChannel = xData.c;
			var p = oTxat.promoteUser(oTxat.getChannelId(sChannel), oTxat.getUserId(sUser), nId);
		} catch (e) {
			txatErrorMessage(oSocket, e);
		}
	});

	oSocket.on(CONST.OPCODE.C_DEMOTE, function(xData) {
		try {
			var nId = wsh.getData(oSocket, 'txat_id');
			var sUser = xData.u;
			var sChannel = xData.c;
			var p = oTxat.demoteUser(oTxat.getChannelId(sChannel), oTxat.getUserId(sUser), nId);
		} catch (e) {
			txatErrorMessage(oSocket, e);
		}
	});

	oSocket.on(CONST.OPCODE.C_UPREF, function(xData) {
		try {
			var nId = wsh.getData(oSocket, 'txat_id');
			var sVariable = xData.v;
			var sValue = xData.w;
			oTxat.setUserPref(nId, sVariable, sValue);
		} catch (e) {
			txatErrorMessage(oSocket, e);
		}
	});
	
	oSocket.on(CONST.OPCODE.C_PASSWD, function(xData) {
		try {
			var nId = wsh.getData(oSocket, 'txat_id');
			var oUser = oTxat.getUser(nId);
			Password.change(oUser.getName(), xData.p);
			oSocket.emit(CONST.OPCODE.IM, {c: null, m: util.format(CONST.MESSAGE.CHANGED_PASSWORD, oUser.getName())});
		} catch (e) {
			txatErrorMessage(oSocket, e);
		}
	});

}



function txatErrorMessage(oSocket, e) {
	oSocket.emit(CONST.OPCODE.ER, {e: e.toString()});
}

function txatGetChannelUserSockets(oChannel) {
	return oChannel.getUserList().map(function(i, n, a) {
		return aSocketRegistry[i];
	});
}


function txatMessage(sMessage, oFromChannel, oFromUser, oToUser, oExtraData) {
	if (oFromChannel) {
		aSocketRegistry[oToUser.id].emit(CONST.OPCODE.CM, {u: oFromUser.getName(), c: oFromChannel.sName, m: sMessage, d: oExtraData});
	} else {
		aSocketRegistry[oToUser.id].emit(CONST.OPCODE.UM, {u: oFromUser.getName(), m: sMessage, d: oExtraData});
	}
}

function txatNotify(sMessage) {
	var oUser, oChannel;
	switch (sMessage) {
		case 'userJoined':
			oChannel = arguments[1];
			oUser = arguments[2];
			wsh.send(txatGetChannelUserSockets(oChannel), CONST.OPCODE.CA, {u: oUser.sName, c: oChannel.sName});
		break;

		case 'userLeft':
			oChannel = arguments[1];
			oUser = arguments[2];
			wsh.send(txatGetChannelUserSockets(oChannel), CONST.OPCODE.CD, {u: oUser.sName, c: oChannel.sName});
		break;
		
		case 'userList':
			oUser = arguments[1];
			oChannel = arguments[2];
			var aUserList = arguments[3];
			wsh.send([aSocketRegistry[oUser.id]], CONST.OPCODE.CL, {c: oChannel.sName, u: aUserList});
		break;
		
		case 'newAdmin':
			oChannel = arguments[1];
			oUser = arguments[2];
			wsh.send(txatGetChannelUserSockets(oChannel), CONST.OPCODE.IM, {c: oChannel.sName, m: util.format(CONST.MESSAGE.NEW_ADMIN, oUser.sName)});
		break;
		
		case 'userAccessDenied':
			oChannel = arguments[1];
			oUser = arguments[2];
			var sReport = oChannel.getBanReport(oUser.sName, '<br />');
			txatErrorMessage(aSocketRegistry[oUser.id], CONST.MESSAGE.ACCESS_DENIED + ' ' + sReport);
		break;

		case 'userBanned':
			oChannel = arguments[1];
			oUser = arguments[2];
			var nTime = arguments[3];
			var sWhy = arguments[4];
			var oJudge = arguments[5];
			var sBannishmentMessage = util.format(CONST.MESSAGE.USER_BANNED, oUser.sName) + oChannel.getBanReport(oUser.sName, ' - ');
			wsh.send(txatGetChannelUserSockets(oChannel), CONST.OPCODE.IM, {c: oChannel.sName, m: sBannishmentMessage});
			txatErrorMessage(aSocketRegistry[oUser.id], util.format(CONST.MESSAGE.BANNED, oChannel.sName) + oChannel.getBanReport(oUser.sName, '<br />'));
			wsh.send(aSocketRegistry[oUser.id], CONST.OPCODE.IM, {c: oChannel.sName, m: sBannishmentMessage});
		break;
		
		case 'userPromote':
		case 'userDemote':
			oChannel = arguments[1];
			oUser = arguments[2];
			wsh.send(txatGetChannelUserSockets(oChannel), CONST.OPCODE.IM, {c: oChannel.sName, m: util.format(CONST.MESSAGE.IS_NOW, oUser.sName, oUser.oPowers[oChannel.id].getRankStr())});
		break;		
	}
}

module.exports = {
	service: txatService,
	init: txatInit
};
