// txat/User.js
var O2 = require('o2');
var ClassMagic = require('classmagic');

var Powers = require('./Powers.js');

/**
 * User : classe gérant un utilisateur
 */
var User = O2.createClass({
	id: 0,				// identifiant de connexion dans le systeme
	sName: '',			// nom de login
	sDisplayName: '',	// Nom affiché
	oPowers: null,		// Liste des droits d'access
	
	__construct: function() {
		this.oPowers = {};
	},
	
	getName: function() {
		return this.sName;
	},
	
	getDisplayName: function() {
		return this.getName();
	},
	
	/**
	 * Fait rejoindre le user dans un canal
	 * Lui octroi des droit standars sauf s'il est admin
	 * @param int idChannel idetifiant canal
	 * @param bool bAsAdmin true = prendre un profil admin
	 */
	grantPowers: function(idChannel, bAsAdmin) {
		var p = new Powers();
		if (this.bSuperAdmin) {
			p.preset(3);
		} else if (bAsAdmin) {
			p.preset(2);
		} else {
			p.preset(0);
		}
		this.oPowers[idChannel] = p;
	},
	
	/**
	 * Supprime les pouvoirs de l'utilisaeur lié au canal spécifié
	 * @param int idChannel
	 */
	stripPowers: function(idChannel) {
		delete this.oPowers[idChannel];
	},
	
	/**
	 * Renvoie la liste des canaux auquel l'utilisateur est connecté
	 * @return array
	 */
	getChannelList: function() {
		return Object.keys(this.oPowers).map(function(i, n, a) {
			return i | 0;
		});
	}
});

ClassMagic.castDataManager(User);
module.exports = User;
